File: web-sdks/javascript/src/core/http_client.js
/**
 * Professional HTTP client implementation using fetch
 *
 * Provides enterprise-grade HTTP functionality with interceptors,
 * retry logic, caching, and comprehensive error handling.
 */

import {
  NetworkException,
  HttpException,
  AuthenticationException,
  RateLimitException,
  AuthErrorType,
} from './exceptions.js';
import { GroupVanResponse, RequestMetadata, ResponseMetadata } from './response.js';
import { GroupVanLogger } from '../logging.js';

/**
 * Generate a UUID v4
 * @returns {string}
 */
function generateUUID() {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
    const r = Math.random() * 16 | 0;
    const v = c === 'x' ? r : (r & 0x3 | 0x8);
    return v.toString(16);
  });
}

/**
 * Configuration for the HTTP client
 */
export class HttpClientConfig {
  /**
   * @param {Object} options
   * @param {string} options.baseUrl - Base URL for API requests
   * @param {number} [options.timeoutMs=30000] - Default timeout for requests in milliseconds
   * @param {number} [options.connectTimeoutMs=10000] - Default timeout for connections
   * @param {number} [options.maxRetries=3] - Maximum number of retry attempts
   * @param {boolean} [options.enableLogging=true] - Enable request/response logging
   * @param {boolean} [options.enableCaching=true] - Enable caching
   * @param {number} [options.cacheDurationMs=300000] - Default cache duration (5 minutes)
   * @param {string} [options.token] - API token for authentication
   * @param {Object} [options.defaultHeaders={}] - Additional default headers
   */
  constructor({
    baseUrl,
    timeoutMs = 30000,
    connectTimeoutMs = 10000,
    maxRetries = 3,
    enableLogging = true,
    enableCaching = true,
    cacheDurationMs = 300000,
    token = null,
    defaultHeaders = {},
  }) {
    this.baseUrl = baseUrl;
    this.timeoutMs = timeoutMs;
    this.connectTimeoutMs = connectTimeoutMs;
    this.maxRetries = maxRetries;
    this.enableLogging = enableLogging;
    this.enableCaching = enableCaching;
    this.cacheDurationMs = cacheDurationMs;
    this.token = token;
    this.defaultHeaders = defaultHeaders;
  }

  /**
   * Create a copy with new values
   * @param {Partial<HttpClientConfig>} overrides
   * @returns {HttpClientConfig}
   */
  copyWith(overrides = {}) {
    return new HttpClientConfig({
      baseUrl: overrides.baseUrl ?? this.baseUrl,
      timeoutMs: overrides.timeoutMs ?? this.timeoutMs,
      connectTimeoutMs: overrides.connectTimeoutMs ?? this.connectTimeoutMs,
      maxRetries: overrides.maxRetries ?? this.maxRetries,
      enableLogging: overrides.enableLogging ?? this.enableLogging,
      enableCaching: overrides.enableCaching ?? this.enableCaching,
      cacheDurationMs: overrides.cacheDurationMs ?? this.cacheDurationMs,
      token: overrides.token ?? this.token,
      defaultHeaders: overrides.defaultHeaders ?? this.defaultHeaders,
    });
  }
}

/**
 * Simple in-memory cache
 */
class MemoryCache {
  constructor() {
    this._cache = new Map();
  }

  /**
   * @param {string} key
   * @param {*} value
   * @param {number} ttlMs - Time to live in milliseconds
   */
  set(key, value, ttlMs) {
    this._cache.set(key, {
      value,
      expiresAt: Date.now() + ttlMs,
    });
  }

  /**
   * @param {string} key
   * @returns {*|null}
   */
  get(key) {
    const entry = this._cache.get(key);
    if (!entry) return null;
    if (Date.now() > entry.expiresAt) {
      this._cache.delete(key);
      return null;
    }
    return entry.value;
  }

  /**
   * @param {string} key
   */
  delete(key) {
    this._cache.delete(key);
  }

  clear() {
    this._cache.clear();
  }
}

/**
 * Professional HTTP client using fetch
 */
export class GroupVanHttpClient {
  /**
   * @param {HttpClientConfig} config
   */
  constructor(config) {
    this._config = config;
    this._cache = new MemoryCache();
    this._token = config.token;
  }

  /**
   * Get the base URL
   * @returns {string}
   */
  get baseUrl() {
    return this._config.baseUrl;
  }

  /**
   * Get the current origin (browser only)
   * @returns {string}
   */
  get origin() {
    if (typeof window !== 'undefined') {
      return window.location.origin;
    }
    return '';
  }

  /**
   * Update the authentication token
   * @param {string|null} token
   */
  updateToken(token) {
    this._token = token;
  }

  /**
   * Build headers for a request
   * @param {Object} [additionalHeaders={}]
   * @returns {Object}
   */
  _buildHeaders(additionalHeaders = {}) {
    const headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      ...this._config.defaultHeaders,
      ...additionalHeaders,
    };

    if (this._token) {
      headers['Authorization'] = `Bearer ${this._token}`;
    }

    return headers;
  }

  /**
   * Make a GET request
   * @template T
   * @param {string} path
   * @param {Object} [options={}]
   * @param {Object} [options.queryParameters]
   * @param {Object} [options.headers]
   * @param {AbortSignal} [options.signal]
   * @param {function(*): T} [options.decoder]
   * @returns {Promise<GroupVanResponse<T>>}
   */
  async get(path, { queryParameters = null, headers = {}, signal = null, decoder = null } = {}) {
    return this._request('GET', path, { queryParameters, headers, signal, decoder });
  }

  /**
   * Make a POST request
   * @template T
   * @param {string} path
   * @param {Object} [options={}]
   * @param {*} [options.data]
   * @param {Object} [options.queryParameters]
   * @param {Object} [options.headers]
   * @param {AbortSignal} [options.signal]
   * @param {function(*): T} [options.decoder]
   * @returns {Promise<GroupVanResponse<T>>}
   */
  async post(path, { data = null, queryParameters = null, headers = {}, signal = null, decoder = null } = {}) {
    return this._request('POST', path, { data, queryParameters, headers, signal, decoder });
  }

  /**
   * Make a PUT request
   * @template T
   * @param {string} path
   * @param {Object} [options={}]
   * @param {*} [options.data]
   * @param {Object} [options.queryParameters]
   * @param {Object} [options.headers]
   * @param {AbortSignal} [options.signal]
   * @param {function(*): T} [options.decoder]
   * @returns {Promise<GroupVanResponse<T>>}
   */
  async put(path, { data = null, queryParameters = null, headers = {}, signal = null, decoder = null } = {}) {
    return this._request('PUT', path, { data, queryParameters, headers, signal, decoder });
  }

  /**
   * Make a PATCH request
   * @template T
   * @param {string} path
   * @param {Object} [options={}]
   * @param {*} [options.data]
   * @param {Object} [options.queryParameters]
   * @param {Object} [options.headers]
   * @param {AbortSignal} [options.signal]
   * @param {function(*): T} [options.decoder]
   * @returns {Promise<GroupVanResponse<T>>}
   */
  async patch(path, { data = null, queryParameters = null, headers = {}, signal = null, decoder = null } = {}) {
    return this._request('PATCH', path, { data, queryParameters, headers, signal, decoder });
  }

  /**
   * Make a DELETE request
   * @template T
   * @param {string} path
   * @param {Object} [options={}]
   * @param {*} [options.data]
   * @param {Object} [options.queryParameters]
   * @param {Object} [options.headers]
   * @param {AbortSignal} [options.signal]
   * @param {function(*): T} [options.decoder]
   * @returns {Promise<GroupVanResponse<T>>}
   */
  async delete(path, { data = null, queryParameters = null, headers = {}, signal = null, decoder = null } = {}) {
    return this._request('DELETE', path, { data, queryParameters, headers, signal, decoder });
  }

  /**
   * Internal request method with retry logic
   * @template T
   * @param {string} method
   * @param {string} path
   * @param {Object} options
   * @returns {Promise<GroupVanResponse<T>>}
   */
  async _request(method, path, { data = null, queryParameters = null, headers = {}, signal = null, decoder = null }) {
    const correlationId = generateUUID();
    const startTime = new Date();

    // Build URL
    let url = `${this._config.baseUrl}${path}`;
    if (queryParameters) {
      const params = new URLSearchParams();
      for (const [key, value] of Object.entries(queryParameters)) {
        if (value !== null && value !== undefined) {
          params.append(key, String(value));
        }
      }
      const queryString = params.toString();
      if (queryString) {
        url += `?${queryString}`;
      }
    }

    // Build headers
    const requestHeaders = this._buildHeaders(headers);
    requestHeaders['X-Correlation-ID'] = correlationId;

    // Log request
    if (this._config.enableLogging) {
      GroupVanLogger.apiClient.fine(`[${correlationId}] ${method} ${url}`);
      if (data) {
        GroupVanLogger.apiClient.finest(`[${correlationId}] Request body:`, data);
      }
    }

    // Retry loop
    let lastError = null;
    for (let attempt = 0; attempt <= this._config.maxRetries; attempt++) {
      try {
        // Create abort controller for timeout
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this._config.timeoutMs);

        const fetchOptions = {
          method,
          headers: requestHeaders,
          signal: signal || controller.signal,
        };

        if (data !== null && ['POST', 'PUT', 'PATCH', 'DELETE'].includes(method)) {
          fetchOptions.body = JSON.stringify(data);
        }

        const response = await fetch(url, fetchOptions);
        clearTimeout(timeoutId);

        const endTime = new Date();
        const durationMs = endTime.getTime() - startTime.getTime();

        // Parse response
        let responseData;
        const contentType = response.headers.get('content-type');
        if (contentType && contentType.includes('application/json')) {
          responseData = await response.json();
        } else {
          responseData = await response.text();
        }

        // Log response
        if (this._config.enableLogging) {
          GroupVanLogger.apiClient.fine(
            `[${correlationId}] Response: ${response.status} (${JSON.stringify(responseData).length} bytes)`
          );
        }

        // Check for errors
        if (!response.ok) {
          throw this._handleHttpError(response, responseData, url, method, correlationId);
        }

        // Decode data if decoder provided
        const decodedData = decoder ? decoder(responseData) : responseData;

        // Build response metadata
        const responseHeaders = {};
        response.headers.forEach((value, key) => {
          responseHeaders[key] = value;
        });

        return new GroupVanResponse({
          data: decodedData,
          statusCode: response.status,
          headers: responseHeaders,
          requestMetadata: new RequestMetadata({
            method,
            url,
            headers: requestHeaders,
            body: data,
            timestamp: startTime,
            timeout: this._config.timeoutMs,
            retryAttempt: attempt,
            correlationId,
          }),
          responseMetadata: new ResponseMetadata({
            timestamp: endTime,
            durationMs,
            sizeBytes: responseHeaders['content-length'] ? parseInt(responseHeaders['content-length']) : null,
            compressed: responseHeaders['content-encoding'] === 'gzip',
            encoding: responseHeaders['content-encoding'],
            server: responseHeaders['server'],
            correlationId,
          }),
          fromCache: false,
        });
      } catch (error) {
        lastError = error;

        // Don't retry if it's a non-retryable error
        if (error instanceof HttpException && !error.isRetryable) {
          throw error;
        }
        if (error instanceof AuthenticationException) {
          throw error;
        }

        // Check if we should retry
        if (attempt < this._config.maxRetries && this._shouldRetry(error)) {
          const delay = this._calculateRetryDelay(attempt);
          if (this._config.enableLogging) {
            GroupVanLogger.apiClient.info(
              `[${correlationId}] Retrying request ${attempt + 1}/${this._config.maxRetries} after ${delay}ms`
            );
          }
          await this._sleep(delay);
          continue;
        }

        throw error;
      }
    }

    throw lastError;
  }

  /**
   * Determine if an error is retryable
   * @param {Error} error
   * @returns {boolean}
   */
  _shouldRetry(error) {
    if (error.name === 'AbortError') {
      return true; // Timeout
    }
    if (error instanceof NetworkException) {
      return true;
    }
    if (error instanceof HttpException) {
      return error.isRetryable;
    }
    if (error instanceof RateLimitException) {
      return true;
    }
    return false;
  }

  /**
   * Calculate retry delay with exponential backoff
   * @param {number} attempt
   * @returns {number} Delay in milliseconds
   */
  _calculateRetryDelay(attempt) {
    const baseDelay = 1000;
    const multiplier = 2;
    return Math.round(baseDelay * Math.pow(multiplier, attempt));
  }

  /**
   * Sleep for a given duration
   * @param {number} ms
   * @returns {Promise<void>}
   */
  _sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Handle HTTP error responses
   * @param {Response} response
   * @param {*} responseData
   * @param {string} url
   * @param {string} method
   * @param {string} correlationId
   * @returns {Error}
   */
  _handleHttpError(response, responseData, url, method, correlationId) {
    const statusCode = response.status;
    const responseBody = typeof responseData === 'string' ? responseData : JSON.stringify(responseData);

    // Handle specific error types
    if (statusCode === 400 && responseData?.title === 'User not found') {
      const email = responseData?.detail?.split(' ')[1] || '';
      return new AuthenticationException(
        'FedLink account must be linked to sign in.',
        {
          errorType: AuthErrorType.ACCOUNT_NOT_LINKED,
          context: { correlationId, email },
        }
      );
    }

    if (statusCode === 401) {
      return new AuthenticationException(
        `Authentication failed: ${responseData?.message || 'Invalid token'}`,
        {
          errorType: AuthErrorType.INVALID_TOKEN,
          context: { correlationId },
        }
      );
    }

    if (statusCode === 403) {
      return new AuthenticationException(
        `Access forbidden: ${responseData?.message || 'Insufficient permissions'}`,
        {
          errorType: AuthErrorType.INSUFFICIENT_PERMISSIONS,
          context: { correlationId },
        }
      );
    }

    if (statusCode === 429) {
      const retryAfter = response.headers.get('retry-after');
      return new RateLimitException(
        `Rate limit exceeded`,
        {
          retryAfterSeconds: retryAfter ? parseInt(retryAfter) : null,
          context: { correlationId },
        }
      );
    }

    return new HttpException(
      responseData?.message || `HTTP error ${statusCode}`,
      {
        statusCode,
        endpoint: url,
        method,
        responseBody,
        context: { correlationId },
      }
    );
  }

  /**
   * Close the HTTP client and clear cache
   */
  close() {
    this._cache.clear();
  }
}

File: web-sdks/dart/lib/src/core/http_client.dart
/// Professional HTTP client implementation using Dio
///
/// Provides enterprise-grade HTTP functionality with interceptors,
/// retry logic, caching, and comprehensive error handling.
library http_client;

import 'dart:convert';

import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:web/web.dart' hide Response;
import 'package:dio/dio.dart';
import 'package:dio/browser.dart';
import 'package:dio_cache_interceptor/dio_cache_interceptor.dart';
import 'package:uuid/uuid.dart';

import '../logging.dart';
import 'exceptions.dart';
import 'response.dart';

/// Configuration for the HTTP client
class HttpClientConfig {
  /// Base URL for API requests
  final String baseUrl;

  /// Default timeout for requests
  final Duration timeout;

  /// Default timeout for connections
  final Duration connectTimeout;

  /// Default timeout for receiving data
  final Duration receiveTimeout;

  /// Maximum number of retry attempts
  final int maxRetries;

  /// Enable request/response logging
  final bool enableLogging;

  /// Enable caching
  final bool enableCaching;

  /// Cache directory path
  final String? cacheDirectory;

  /// Default cache duration
  final Duration cacheDuration;

  /// API token for authentication
  final String? token;

  /// Additional default headers
  final Map<String, String> defaultHeaders;

  const HttpClientConfig({
    required this.baseUrl,
    this.timeout = const Duration(seconds: 30),
    this.connectTimeout = const Duration(seconds: 10),
    this.receiveTimeout = const Duration(seconds: 30),
    this.maxRetries = 3,
    this.enableLogging = true,
    this.enableCaching = true,
    this.cacheDirectory,
    this.cacheDuration = const Duration(minutes: 5),
    this.token,
    this.defaultHeaders = const {},
  });

  HttpClientConfig copyWith({
    String? baseUrl,
    Duration? timeout,
    Duration? connectTimeout,
    Duration? receiveTimeout,
    int? maxRetries,
    bool? enableLogging,
    bool? enableCaching,
    String? cacheDirectory,
    Duration? cacheDuration,
    String? token,
    Map<String, String>? defaultHeaders,
  }) {
    return HttpClientConfig(
      baseUrl: baseUrl ?? this.baseUrl,
      timeout: timeout ?? this.timeout,
      connectTimeout: connectTimeout ?? this.connectTimeout,
      receiveTimeout: receiveTimeout ?? this.receiveTimeout,
      maxRetries: maxRetries ?? this.maxRetries,
      enableLogging: enableLogging ?? this.enableLogging,
      enableCaching: enableCaching ?? this.enableCaching,
      cacheDirectory: cacheDirectory ?? this.cacheDirectory,
      cacheDuration: cacheDuration ?? this.cacheDuration,
      token: token ?? this.token,
      defaultHeaders: defaultHeaders ?? this.defaultHeaders,
    );
  }
}

/// Professional HTTP client using Dio
class GroupVanHttpClient {
  late final Dio _dio;
  final HttpClientConfig _config;
  final Uuid _uuid = const Uuid();

  /// Cache store for HTTP responses
  CacheStore? _cacheStore;

  GroupVanHttpClient(this._config) {
    _initializeDio();
  }

  /// Initialize Dio with all interceptors and configuration
  void _initializeDio() {
    _dio = Dio(
      BaseOptions(
        baseUrl: _config.baseUrl,
        connectTimeout: _config.connectTimeout,
        receiveTimeout: _config.receiveTimeout,
        sendTimeout: _config.timeout,
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          if (_config.token != null) 'Authorization': 'Bearer ${_config.token}',
          ..._config.defaultHeaders,
        },
      ),
    );

    // On web, enable withCredentials so the browser includes HttpOnly cookies
    // (refresh_token) in cross-origin requests to *.groupvan.com
    if (kIsWeb) {
      _dio.httpClientAdapter = BrowserHttpClientAdapter(withCredentials: true);
    }

    // Sanitize sendTimeout for requests without a body (esp. required on Web)
    _dio.interceptors.add(SendTimeoutSanitizerInterceptor());

    // Add authentication interceptor
    _dio.interceptors.add(AuthInterceptor(_config.token));

    // Add retry interceptor
    _dio.interceptors.add(RetryInterceptor(_config.maxRetries));

    // Add caching if enabled
    if (_config.enableCaching) {
      _initializeCache();
    }

    // Add logging interceptor (should be last for complete request/response logging)
    if (_config.enableLogging) {
      _dio.interceptors.add(LoggingInterceptor());
    }

    // Add correlation ID interceptor
    _dio.interceptors.add(CorrelationInterceptor(_uuid));

    // Add error handling interceptor
    _dio.interceptors.add(ErrorHandlingInterceptor());
  }

  /// Initialize cache store
  void _initializeCache() {
    final cacheOptions = CacheOptions(
      store: _cacheStore ?? MemCacheStore(),
      policy: CachePolicy.request,
      // hitCacheOnErrorExcept removed for compatibility
      maxStale: const Duration(days: 7),
      priority: CachePriority.normal,
      cipher: null,
      allowPostMethod: false,
    );

    _dio.interceptors.add(DioCacheInterceptor(options: cacheOptions));
  }

  String get baseUrl => _config.baseUrl;

  String get origin => window.location.origin;

  /// Make a GET request
  Future<GroupVanResponse<T>> get<T>(
    String path, {
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    T Function(dynamic)? decoder,
  }) async {
    final startTime = DateTime.now();
    final correlationId = _uuid.v4();

    try {
      final response = await _dio.get<dynamic>(
        path,
        queryParameters: queryParameters,
        options:
            options?.copyWith(extra: {'correlation_id': correlationId}) ??
            Options(extra: {'correlation_id': correlationId}),
        cancelToken: cancelToken,
      );

      return _buildResponse<T>(
        response,
        startTime,
        correlationId,
        decoder: decoder,
      );
    } on DioException catch (e) {
      throw _handleDioException(e, correlationId);
    }
  }

  /// Make a POST request
  Future<GroupVanResponse<T>> post<T>(
    String path, {
    dynamic data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    T Function(dynamic)? decoder,
  }) async {
    final startTime = DateTime.now();
    final correlationId = _uuid.v4();

    try {
      final response = await _dio.post<dynamic>(
        path,
        data: data,
        queryParameters: queryParameters,
        options:
            options?.copyWith(extra: {'correlation_id': correlationId}) ??
            Options(extra: {'correlation_id': correlationId}),
        cancelToken: cancelToken,
      );

      return _buildResponse<T>(
        response,
        startTime,
        correlationId,
        decoder: decoder,
      );
    } on DioException catch (e) {
      throw _handleDioException(e, correlationId);
    }
  }

  /// Make a PUT request
  Future<GroupVanResponse<T>> put<T>(
    String path, {
    dynamic data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    T Function(dynamic)? decoder,
  }) async {
    final startTime = DateTime.now();
    final correlationId = _uuid.v4();

    try {
      final response = await _dio.put<dynamic>(
        path,
        data: data,
        queryParameters: queryParameters,
        options:
            options?.copyWith(extra: {'correlation_id': correlationId}) ??
            Options(extra: {'correlation_id': correlationId}),
        cancelToken: cancelToken,
      );

      return _buildResponse<T>(
        response,
        startTime,
        correlationId,
        decoder: decoder,
      );
    } on DioException catch (e) {
      throw _handleDioException(e, correlationId);
    }
  }

  /// Make a PATCH request
  Future<GroupVanResponse<T>> patch<T>(
    String path, {
    dynamic data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    T Function(dynamic)? decoder,
  }) async {
    final startTime = DateTime.now();
    final correlationId = _uuid.v4();

    try {
      final response = await _dio.patch<dynamic>(
        path,
        data: data,
        queryParameters: queryParameters,
        options:
            options?.copyWith(extra: {'correlation_id': correlationId}) ??
            Options(extra: {'correlation_id': correlationId}),
        cancelToken: cancelToken,
      );

      return _buildResponse<T>(
        response,
        startTime,
        correlationId,
        decoder: decoder,
      );
    } on DioException catch (e) {
      throw _handleDioException(e, correlationId);
    }
  }

  /// Make a DELETE request
  Future<GroupVanResponse<T>> delete<T>(
    String path, {
    dynamic data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
    T Function(dynamic)? decoder,
  }) async {
    final startTime = DateTime.now();
    final correlationId = _uuid.v4();

    try {
      final response = await _dio.delete<dynamic>(
        path,
        data: data,
        queryParameters: queryParameters,
        options:
            options?.copyWith(extra: {'correlation_id': correlationId}) ??
            Options(extra: {'correlation_id': correlationId}),
        cancelToken: cancelToken,
      );

      return _buildResponse<T>(
        response,
        startTime,
        correlationId,
        decoder: decoder,
      );
    } on DioException catch (e) {
      throw _handleDioException(e, correlationId);
    }
  }

  /// Build GroupVanResponse from Dio response
  GroupVanResponse<T> _buildResponse<T>(
    Response<dynamic> response,
    DateTime startTime,
    String correlationId, {
    T Function(dynamic)? decoder,
  }) {
    final endTime = DateTime.now();
    final duration = endTime.difference(startTime);

    // Decode data using provided decoder or return as-is
    final T data = decoder != null
        ? decoder(response.data)
        : response.data as T;

    // Check if response was cached
    final fromCache =
        response.headers.value('cache-control')?.contains('hit') ?? false;
    final cacheDate = response.headers.value('date');
    final cacheTimestamp = cacheDate != null
        ? DateTime.tryParse(cacheDate)
        : null;

    return GroupVanResponse<T>(
      data: data,
      statusCode: response.statusCode ?? 0,
      headers: response.headers.map.map(
        (key, value) => MapEntry(key, value.join(', ')),
      ),
      requestMetadata: RequestMetadata(
        method: response.requestOptions.method,
        url: response.realUri.toString(),
        headers: response.requestOptions.headers.cast<String, String>(),
        body: response.requestOptions.data,
        timestamp: startTime,
        timeout:
            response.requestOptions.sendTimeout ?? const Duration(seconds: 30),
        retryAttempt: response.requestOptions.extra['retry_count'] ?? 0,
        correlationId: correlationId,
      ),
      responseMetadata: ResponseMetadata(
        timestamp: endTime,
        duration: duration,
        sizeBytes: response.headers.value('content-length') != null
            ? int.tryParse(response.headers.value('content-length')!)
            : null,
        compressed: response.headers.value('content-encoding') == 'gzip',
        encoding: response.headers.value('content-encoding'),
        server: response.headers.value('server'),
        correlationId: correlationId,
      ),
      fromCache: fromCache,
      cacheTimestamp: cacheTimestamp,
    );
  }

  /// Handle DioException and convert to GroupVan exception
  GroupVanException _handleDioException(DioException e, String correlationId) {
    switch (e.type) {
      case DioExceptionType.connectionTimeout:
      case DioExceptionType.sendTimeout:
      case DioExceptionType.receiveTimeout:
        return NetworkException(
          'Request timeout: ${e.message}',
          endpoint: e.requestOptions.path,
          method: e.requestOptions.method,
          context: {'correlation_id': correlationId},
        );

      case DioExceptionType.badResponse:
        final statusCode = e.response?.statusCode ?? 0;
        final responseData = e.response?.data?.toString();

        if (statusCode == 400 && responseData != null) {
          Map<String, dynamic> responseJson = e.response!.data;
          if (responseJson['title'] == 'User not found') {
            final email = responseJson['detail'].split(' ')[1];
            return AuthenticationException(
              'FedLink account must be linked to sign in.',
              errorType: AuthErrorType.accountNotLinked,
              context: {'correlation_id': correlationId, 'email': email},
            );
          }
        }

        // Handle authentication errors
        if (statusCode == 401) {
          return AuthenticationException(
            'Authentication failed: ${e.message}',
            errorType: AuthErrorType.invalidToken,
            context: {'correlation_id': correlationId},
          );
        }

        // Handle authorization errors
        if (statusCode == 403) {
          return AuthenticationException(
            'Access forbidden: ${e.message}',
            errorType: AuthErrorType.insufficientPermissions,
            context: {'correlation_id': correlationId},
          );
        }

        // Handle rate limiting
        if (statusCode == 429) {
          final retryAfter = e.response?.headers.value('retry-after');
          return RateLimitException(
            'Rate limit exceeded: ${e.message}',
            retryAfterSeconds: retryAfter != null
                ? int.tryParse(retryAfter)
                : null,
            context: {'correlation_id': correlationId},
          );
        }

        return HttpException(
          e.message ?? 'HTTP error occurred',
          statusCode: statusCode,
          endpoint: e.requestOptions.path,
          method: e.requestOptions.method,
          responseBody: responseData,
          context: {'correlation_id': correlationId},
        );

      case DioExceptionType.cancel:
        return NetworkException(
          'Request was cancelled',
          endpoint: e.requestOptions.path,
          method: e.requestOptions.method,
          context: {'correlation_id': correlationId},
        );

      case DioExceptionType.connectionError:
        return NetworkException(
          'Connection error: ${e.message}',
          endpoint: e.requestOptions.path,
          method: e.requestOptions.method,
          context: {'correlation_id': correlationId},
        );

      case DioExceptionType.badCertificate:
        return NetworkException(
          'SSL certificate error: ${e.message}',
          endpoint: e.requestOptions.path,
          method: e.requestOptions.method,
          context: {'correlation_id': correlationId},
        );

      case DioExceptionType.unknown:
        return NetworkException(
          'Unknown error: ${e.message}',
          endpoint: e.requestOptions.path,
          method: e.requestOptions.method,
          context: {'correlation_id': correlationId},
        );
    }
  }

  /// Close the HTTP client
  void close() {
    _dio.close();
  }
}

/// Send-timeout sanitizer to avoid using sendTimeout on requests without a body
class SendTimeoutSanitizerInterceptor extends Interceptor {
  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) {
    final method = options.method.toUpperCase();
    final hasNoBody = options.data == null;

    if (hasNoBody &&
        (method == 'GET' || method == 'HEAD' || method == 'OPTIONS')) {
      options.sendTimeout = null;
    }

    handler.next(options);
  }
}

/// Authentication interceptor
class AuthInterceptor extends Interceptor {
  String? _token;

  AuthInterceptor(this._token);

  void updateToken(String? token) {
    _token = token;
  }

  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) {
    if (_token != null) {
      options.headers['Authorization'] = 'Bearer $_token';
    }
    handler.next(options);
  }
}

/// Retry interceptor with exponential backoff
class RetryInterceptor extends Interceptor {
  final int maxRetries;
  final Duration baseDelay;
  final double backoffMultiplier;

  RetryInterceptor(
    this.maxRetries, {
    this.baseDelay = const Duration(seconds: 1),
    this.backoffMultiplier = 2.0,
  });

  @override
  void onError(DioException err, ErrorInterceptorHandler handler) async {
    final retryCount = err.requestOptions.extra['retry_count'] ?? 0;

    if (_shouldRetry(err) && retryCount < maxRetries) {
      // Calculate delay with exponential backoff
      final delay = Duration(
        milliseconds:
            (baseDelay.inMilliseconds * (1 << retryCount) * backoffMultiplier)
                .round(),
      );

      GroupVanLogger.apiClient.info(
        'Retrying request ${retryCount + 1}/$maxRetries after ${delay.inMilliseconds}ms',
      );

      await Future.delayed(delay);

      // Update retry count
      err.requestOptions.extra['retry_count'] = retryCount + 1;

      // Retry the request
      try {
        final response = await Dio().fetch(err.requestOptions);
        handler.resolve(response);
      } on DioException catch (e) {
        handler.next(e);
      }
    } else {
      handler.next(err);
    }
  }

  bool _shouldRetry(DioException error) {
    // Retry on network errors and 5xx server errors
    return error.type == DioExceptionType.connectionTimeout ||
        error.type == DioExceptionType.sendTimeout ||
        error.type == DioExceptionType.receiveTimeout ||
        error.type == DioExceptionType.connectionError ||
        (error.response?.statusCode != null &&
            error.response!.statusCode! >= 500) ||
        error.response?.statusCode == 429; // Rate limit
  }
}

/// Logging interceptor
class LoggingInterceptor extends Interceptor {
  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) {
    final correlationId = options.extra['correlation_id'] ?? 'unknown';
    GroupVanLogger.apiClient.fine(
      '[$correlationId] ${options.method} ${options.uri}',
    );

    if (options.data != null) {
      GroupVanLogger.apiClient.finest(
        '[$correlationId] Request body: ${options.data}',
      );
    }

    handler.next(options);
  }

  @override
  void onResponse(Response response, ResponseInterceptorHandler handler) {
    final correlationId =
        response.requestOptions.extra['correlation_id'] ?? 'unknown';
    GroupVanLogger.apiClient.fine(
      '[$correlationId] Response: ${response.statusCode} (${response.data?.toString().length ?? 0} bytes)',
    );

    handler.next(response);
  }

  @override
  void onError(DioException err, ErrorInterceptorHandler handler) {
    final correlationId =
        err.requestOptions.extra['correlation_id'] ?? 'unknown';
    GroupVanLogger.apiClient.warning(
      '[$correlationId] Error: ${err.type} - ${err.message}',
    );

    handler.next(err);
  }
}

/// Correlation ID interceptor
class CorrelationInterceptor extends Interceptor {
  final Uuid _uuid;

  CorrelationInterceptor(this._uuid);

  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) {
    // Add correlation ID if not already present
    if (!options.extra.containsKey('correlation_id')) {
      options.extra['correlation_id'] = _uuid.v4();
    }

    // Add correlation ID to headers
    options.headers['X-Correlation-ID'] = options.extra['correlation_id'];

    handler.next(options);
  }
}

/// Error handling interceptor
class ErrorHandlingInterceptor extends Interceptor {
  @override
  void onError(DioException err, ErrorInterceptorHandler handler) {
    // Log error details
    final correlationId =
        err.requestOptions.extra['correlation_id'] ?? 'unknown';
    GroupVanLogger.apiClient.severe(
      '[$correlationId] HTTP Error: ${err.type} - ${err.message}',
    );

    handler.next(err);
  }
}
